pkg/types/keyval.go

```diff
@@ -0,0 +1,313 @@
+// no additional information
+func (k Key) ExtendEmpty() (e ExtendedKey) {
+	return k.Extend(0, "")
+}
+
+// ExtendedKey is a Key with supplemental information
+type ExtendedKey []byte
+
+// Key retrieves the basic key within the extended key to allow for
+// more precise access without having to always use the (longer) ExtendedKey
+func (e ExtendedKey) Key() Key {
+	if e.IsIPv4() {
+		return Key(e[1 : 1+KeyWidthIPv4])
+	}
+	return Key(e[1 : 1+KeyWidthIPv6])
+}
+
+// IsIPv4 returns if the key represents an IPv4 packet / flow
+func (e ExtendedKey) IsIPv4() bool {
+	return e[0]&(1<<0) <= 0
+}
+
+// PutSip stores a source IP in the key
+func (e ExtendedKey) PutSip(sip []byte) {
+	copy(e[1+SipPos:], sip)
+}
+
+// PutSipV4 stores an IPv4 source IP in the key
+func (e ExtendedKey) PutSipV4(sip []byte) {
+	copy(e[1+SipPos:1+SipPos+IPv4Width], sip)
+}
+
+// PutSipV6 stores an IPv6 source IP in the key
+func (e ExtendedKey) PutSipV6(sip []byte) {
+	copy(e[1+SipPos:1+SipPos+IPv6Width], sip)
+}
+
+// PutDip stores a destination IP in the key
+func (e ExtendedKey) PutDip(dip []byte) {
+	if len(dip) == IPv4Width {
+		copy(e[1+DipPosIPv4:], dip)
+	} else {
+		copy(e[1+DipPosIPv6:], dip)
+	}
+}
+
+// PutDipV4 stores an IPv4 destination IP in the key
+func (e ExtendedKey) PutDipV4(dip []byte) {
+	copy(e[1+DipPosIPv4:1+DipPosIPv4+IPv4Width], dip)
+}
+
+// PutDipV6 stores an IPv6 destination IP in the key
+func (e ExtendedKey) PutDipV6(dip []byte) {
+	copy(e[1+DipPosIPv6:1+DipPosIPv6+IPv6Width], dip)
+}
+
+// PutDport stores a destination port in the key
+func (e ExtendedKey) PutDport(dport []byte) {
+	copy(e[1+DPortPos:1+DPortPos+DPortWidth], dport[:])
+}
+
+// PutProto stores a protocol in the key
+func (e ExtendedKey) PutProto(proto byte) {
+	copy(e[1+ProtoPos:1+ProtoPos+ProtoWidth], []byte{proto})
+}
+
+// AttrTime retrieves the time extension (indicating its presence via the second result parameter)
+func (e ExtendedKey) AttrTime() (int64, bool) {
+	if e[0]&(1<<1) > 0 {
+		if e.IsIPv4() {
+			return int64(binary.BigEndian.Uint64(e[1+KeyWidthIPv4 : 9+KeyWidthIPv4])), true
+		}
+		return int64(binary.BigEndian.Uint64(e[1+KeyWidthIPv6 : 9+KeyWidthIPv6])), true
+	}
+	return 0, false
+}
+
+// AttrIface retrieves the interface name extension (indicating its presence via the second result parameter)
+func (e ExtendedKey) AttrIface() (string, bool) {
+	if e[0]&(1<<2) > 0 {
+		pos := 1
+		if e.IsIPv4() {
+			pos += KeyWidthIPv4
+		} else {
+			pos += KeyWidthIPv6
+		}
+		if e[0]&(1<<1) > 0 {
+			pos += 8
+		}
+		return string(e[pos:]), true
+	}
+	return "", false
+}
+
+// String prints the key as a comma separated attribute list
+func (k Key) String() string {
+	return fmt.Sprintf("%s,%s,%d,%s",
+		RawIPToString(k.GetSip()),
+		RawIPToString(k.GetDip()),
+		int(PortToUint16(k.GetDport())),
+		protocols.GetIPProto(int(k.GetProto())),
+	)
+}
+
+// Val stores the goProbe flow counters (and, where required, some extensions)
+type Val struct {
```

Comment:
Now that I see it: the data structure is identical to `results.Counters`. Why don't we use that instead?

Reply:
Fair point. I've also added an `func (c Counters) Add(c2 Counters) Counters` method to allow for a cleaner way of appending to these counters without always having to manually do:
```
       c.NBytesRcvd += c2.NBytesRcvd
       c.NBytesSent += c2.NBytesSent
       c.NPktsRcvd += c2.NPktsRcvd
       c.NPktsSent += c2.NPktsSent
```